openmrs/omod/src/main/java/org/openmrs/projectbuendia/webservices/rest/EncounterResource.java

```diff
@@ -39,43 +40,34 @@
  * point in time (referred to as the "snapshot time").
  * @see AbstractReadOnlyResource
  */
-// TODO: Merge with PatientResource and let clients use a query parameter to
-// indicate whether encounter data should be included with each returned patient.
 @Resource(name = RestController.REST_VERSION_1_AND_NAMESPACE + "/encounters",
     supportedClass = Patient.class, supportedOpenmrsVersions = "1.10.*,1.11.*")
 public class EncounterResource
-    extends AbstractReadOnlyResource<Patient> implements Creatable {
+    extends AbstractReadOnlyResource<Encounter> implements Creatable {
     private final PatientService patientService;
-    private final EncounterService encounterService;
+    private final ProjectBuendiaService buendiaService;
 
     public EncounterResource() {
-        super("patient", Representation.DEFAULT);
+        super("encounter", Representation.DEFAULT);
         patientService = Context.getPatientService();
-        encounterService = Context.getEncounterService();
+        buendiaService = Context.getService(ProjectBuendiaService.class);
     }
 
     /**
-     * Returns all patients.  The retrieved records will be filled in with
-     * each patient's encounter and observation data by
-     * {@link #populateJsonProperties(Patient, RequestContext, SimpleObject, long)}
-     * on its way to becoming JSON that is sent to the client.
-     * @param context      unused here; see populateJsonProperties() for details
+     * Returns all encounters observed after the "since" parameter, or since the beginning of time
+     * if the "since" parameter is not set.
+     * @param context      used to obtain the "since" parameter for incremental sync.
      * @param snapshotTime unused here; see populateJsonProperties() for details
      * @see AbstractReadOnlyResource#search(RequestContext)
      */
-    @Override public List<Patient> searchImpl(RequestContext context, long snapshotTime) {
-        String patientUuid = context.getParameter("patient_uuid");
-        if (patientUuid != null) {
-            Patient patient = patientService.getPatientByUuid(patientUuid);
-            if (patient == null) {
-                throw new ObjectNotFoundException();
-            }
-            List<Patient> patients = new ArrayList<>();
-            patients.add(patient);
-            return patients;
-        } else {
-            return patientService.getAllPatients();
+    @Override public List<Encounter> searchImpl(RequestContext context, long snapshotTime) {
+        Date syncFrom;
+        try {
+            syncFrom = RequestUtil.getSyncFromDate(context);
```

Comment:
Agreed on all points! :)
On Nov 2, 2015 11:00 PM, "Fabian Tamp" <notifications@github.com> wrote:

> In
> openmrs/omod/src/main/java/org/openmrs/projectbuendia/webservices/rest/EncounterResource.java
> <https://github.com/projectbuendia/buendia/pull/20#discussion_r43717892>:
>
> > -    @Override public List<Patient> searchImpl(RequestContext context, long snapshotTime) {
> > -        String patientUuid = context.getParameter("patient_uuid");
> > -        if (patientUuid != null) {
> > -            Patient patient = patientService.getPatientByUuid(patientUuid);
> > -            if (patient == null) {
> > -                throw new ObjectNotFoundException();
> > -            }
> > -            List<Patient> patients = new ArrayList<>();
> > -            patients.add(patient);
> > -            return patients;
> > -        } else {
> > -            return patientService.getAllPatients();
> > +    @Override public List<Encounter> searchImpl(RequestContext context, long snapshotTime) {
> > +        Date syncFrom;
> > +        try {
> > +            syncFrom = RequestUtil.getSyncFromDate(context);
>
> Ok. I spent another few hours thinking about this this morning; I'm
> convinced about using timestamp + UUID now :)
>
> A few observations:
>
>    - The biggest strength of [timestamp + UUID] is that it means that
>    every HTTP request + apply to local DB results in a determinate state - we
>    don't ever need to rollback transactions if a sync only partially
>    completes. That's extremely powerful, and I think worth having.
>    - I think neither bookmarking method ([timestamp + UUID], [timestamp +
>    page number]) solves the problem where we could miss records added at the
>    same time as the final page is synced. This is something that we have to
>    address by choosing the bookmark returned for the last page of results. I
>    have a proposal for this, I'll add it to the sync doc.
>    - The question of whether to report num_pages at all is slightly
>    orthogonal to the choice of bookmark. It might still be worth providing an
>    estimated number of pages to the client in each request so that they can
>    indicate sync progress, but we can do that in parallel/as an enhancement
>    now.
>
> One meta-observation:
>
>    - I think it probably would've been much quicker if we had've hopped
>    online and chatted about this in person rather than debating it on the
>    thread :) Next time let's chat about this in person.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/projectbuendia/buendia/pull/20/files#r43717892>.
>